File: crates/ffi/lang/node/examples/opencv_demo/emit_manifest.mjs
/**
 * Emit a manifest that demonstrates passing an image-like payload into Node, processing it with
 * OpenCV, and returning it back to Rust.
 *
 * Notes:
 * - Requires an OpenCV binding such as `opencv4nodejs` available at runtime.
 * - The image carrier is a struct `{data_b64,width,height,channels,dtype,layout,encoding}` where
 *   `data_b64` carries bytes; when `encoding=="raw"`, it is raw pixels (HWC, u8).
 */
import os from "node:os";
import path from "node:path";
import { fileURLToPath } from "node:url";

import { Plugin, node, outputs, port, t } from "../../daedalus_node/index.mjs";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const outPath = process.argv[2]
  ? path.resolve(process.argv[2])
  : path.join(os.tmpdir(), `demo_node_opencv_${process.pid}.manifest.json`);

const imageTy = t.struct([
  { name: "data_b64", ty: t.string() },
  { name: "width", ty: t.int() },
  { name: "height", ty: t.int() },
  { name: "channels", ty: t.int() },
  { name: "dtype", ty: t.string() },
  { name: "layout", ty: t.string() },
  { name: "encoding", ty: t.string() },
]);

const modulePath = path.join(__dirname, "runtime.mjs");

export const blur = node({
  id: "demo_node_opencv:blur",
  label: "OpenCV Blur",
  path: modulePath,
  func: "blur",
  inputs: [port("img", imageTy)],
  outputs: outputs(port("out", imageTy)),
})(function blur_impl(_img) {});

const plugin = new Plugin({
  name: "demo_node_opencv",
  version: "0.1.0",
  description: "OpenCV image demo",
});
plugin.register(blur.__daedalus_node__);
plugin.emitManifest(outPath);
console.log(outPath);

File: crates/ffi/lang/python/examples/opencv_demo/emit_manifest.py
"""
Emit a manifest that demonstrates passing an image-like payload into Python, processing it with
OpenCV, and returning it back to Rust.

Notes:
- Requires `opencv-python` at runtime (cv2 import).
- The cross-language image carrier is a small struct with base64 bytes; prefer `encoding=="raw"`.
"""

from __future__ import annotations

import importlib.util
import os
import sys
import tempfile
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(ROOT))

if __name__ == "__main__":
    out = Path(tempfile.gettempdir()) / f"demo_py_opencv_{os.getpid()}.manifest.json"
    if len(sys.argv) > 1:
        out = Path(sys.argv[1])
    out = out.resolve()

    defs_path = Path(__file__).with_name("defs.py")
    spec = importlib.util.spec_from_file_location("demo_py_opencv_defs", defs_path)
    assert spec and spec.loader
    mod = importlib.util.module_from_spec(spec)
    sys.modules[spec.name] = mod
    spec.loader.exec_module(mod)
    mod.plugin.emit_manifest(out)
    print(out.as_posix())
